server/lib/api/parties.ts

```diff
@@ -0,0 +1,163 @@
+import users from '../models/users'
+import PartyService, { PartyServiceError, PartyUser } from '../parties/party-service'
+import ensureLoggedIn from '../session/ensure-logged-in'
+import createThrottle from '../throttle/create-throttle'
+import throttleMiddleware from '../throttle/middleware'
+import { joiValidator } from '../validation/joi-validator'
+
+const partyService = container.resolve(PartyService)
+
+const invitesThrottle = createThrottle('partyInvites', {
+  rate: 40,
+  burst: 60,
+  window: 60000,
+})
+
+const partyThrottle = createThrottle('parties', {
+  rate: 20,
+  burst: 40,
+  window: 60000,
+})
+
+const clientIdSchema = Joi.object().keys({
+  clientId: Joi.string().required(),
+})
+
+const invitesPostchema = clientIdSchema.keys({
+  targets: Joi.array()
+    .items(
+      Joi.string()
+        .min(USERNAME_MINLENGTH)
+        .max(USERNAME_MAXLENGTH)
+        .pattern(USERNAME_PATTERN)
+        .required(),
+    )
+    .min(1)
+    .required(),
+})
+
+const partyIdSchema = Joi.object().keys({
+  partyId: Joi.string().required(),
+})
+
+const declineParamsSchema = partyIdSchema.keys({
+  targetId: Joi.number(),
+})
+
+export default function (router: Router) {
+  router
+    .use(ensureLoggedIn)
+    .post(
+      '/invites',
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ body: invitesPostchema }),
+      invite,
+    )
+    .delete(
+      '/invites/:partyId/:targetId',
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: declineParamsSchema }),
+      decline,
+    )
+    .post(
+      '/:partyId',
+      throttleMiddleware(partyThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: partyIdSchema }),
+      accept,
+    )
+}
+
+function isPartyServiceError(error: Error): error is PartyServiceError {
+  return error.hasOwnProperty('code')
+}
+
+async function invite(ctx: RouterContext) {
+  const { clientId, targets } = ctx.request.body
+
+  try {
+    const invites = await Promise.all(
+      targets.map(
+        async (target: string): Promise<PartyUser> => {
+          const foundTarget = await users.find(target)
+          if (!foundTarget) {
+            throw new httpErrors.NotFound('Target user not found')
+          }
+
+          // TODO(2Pac): Check if the target user has blocked invitations from the user issuing the
+          // request. Or potentially use friends list when implemented.
+
+          return new PartyUser({ id: foundTarget.id, name: foundTarget.name })
+        },
+      ),
+    )
+
+    const leader = new PartyUser({
+      id: ctx.session!.userId,
+      name: ctx.session!.userName,
+    })
+
+    partyService.invite(leader, clientId, invites as PartyUser[])
+
+    ctx.status = 204
+  } catch (err) {
+    if (isPartyServiceError(err)) {
+      throw new httpErrors[err.code](err.message)
```

Comment:
So if you're going this route, you may as well just throw http-errors from the PartyService, that's not really what I was suggesting. What I meant is that the PartyService could throw error codes specific to it (its own enum of possible error conditions) and you'd have a function here that would convert a PartyServiceError code into the relevant http one

This would allow the PartyService to exist separate from whatever transport it ends up being on, and the tests for it to make sense without any understanding of HTTP